buckets-mdapi/src/object/get.rs

```diff
@@ -47,14 +48,25 @@ pub(crate) fn action(
             Ok(msg)
         })
         .or_else(|e| {
-            // Handle database error response
-            error!(log, "operation failed"; "error" => &e);
+            /*
+             * At this point we've seen some kind of failure processing the request.  It could be
+             * that the database has returned an error of some kind, or that there has been a
+             * failure in evaluating the conditions of the request.  Either way they are handed
+             * back to the application as fast messages.
+             */
+
+            /*
+             * XXX
+             *
+             * Is reaching into the Value like this safe?  I think we get `Value::Null` on a bad
+             * value, so possibly it's ok.  Is it fast enough?
+             */
+            if e["name"] == "PostgresError" {
```

Comment:
I think we could possibly improve on this a bit and not have to deal with the raw `Value` here. What I'm thinking is maybe we can have `do_get` return a `Result<Option<ObjectResponse>, BucketsMdapiErrorType>`  rather than `Result<Option<ObjectResponse>, Value>`. If we extend `BucketsMdapiErrorTyep` with a new variant that looks like `PostgresError(String)` maybe we can achieve similar results without the need to muck directly with that `Value`. This is pretty hand wavy at this point so if you want I can try this out and share more details about what I mean if it seems like it will work.

Reply:
This one ended up being a bit of a mission! I agree though because the error type was a little bit confusing for me at first.

Basically I've made `enum BucketsMdapiErrorType` into our internal representation of an error, and it's now `enum BucketsMdapiError`. Like you've said, we've now got `String` variants that include a message with the error which is nice. Only when it comes to serializing and deserializing do we wrap this into `BucketsMdapiWrappedError`, which buys us some niceties from Serde with wrapping it up into what Fast expects our errors to look like. I've kind of just inverted the existing `BucketsMdapiErrorType` and `BucketsMdapiError` types.

I hope that makes sense. It took me a while to get to this point because Serde can _almost_ afford us to do away with `ToString` (except for our test suite) and `BucketsMdapiWrappedError`. I've written that up into a comment.

The delta for just this change is between https://github.com/joyent/manta-buckets-mdapi/pull/47/commits/fe7e575d15bc9def62a033529ec322712020153e and https://github.com/joyent/manta-buckets-mdapi/pull/47/commits/1ceb15bfcaf5e609426bd67949d5801146ac570b.

Again, I think this is doing what I expect, but I'd appreciate extra eyes on it.